src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/BalanceCalculator.java

```diff
@@ -0,0 +1,167 @@
+    this.restClient = restClient;
+  }
+
+  public List<Accrual> calculate(TimeEntry timeEntry) {
+
+    List<Accrual> accrualsToBatchUpdate = Arrays.stream(AccrualType.values())
+        .filter(AccrualType::isEnabled)
+        .map(accrualType -> calculateAccruals(timeEntry, accrualType))
+        .flatMap(Collection::stream)
+        .toList();
+
+    // TODO : send Batch Update request to Accruals API
+    // restClient.batchUpdate(accrualsToBatchUpdate)
+
+    return accrualsToBatchUpdate;
+  }
+
+  List<Accrual> calculateAccruals(TimeEntry timeEntry, AccrualType accrualType) {
+
+    Map<LocalDate, Range<ZonedDateTime>> dateRangeMap =
+        splitOverDays(timeEntry.getActualStartTime(), timeEntry.getActualEndTime());
+
+    return dateRangeMap.entrySet().stream()
+        .map(entry -> recalculateContributions(timeEntry.getId(), timeEntry.getTenantId(),
+            timeEntry.getOwnerId(), accrualType, entry.getKey(), entry.getValue()))
+        .map(accrual -> cascadeCumulativeTotal(timeEntry, accrual)).flatMap(Collection::stream)
+        .toList();
+  }
+
+  private List<Accrual> cascadeCumulativeTotal(TimeEntry timeEntry, Accrual accrual) {
+    LocalDate referenceDate = accrual.getAccrualDate();
+
+    List<Accrual> accrualsToUpdate =
+        getAccrualsFromReferenceDateUntilEndOfAgreement(timeEntry.getTenantId(),
+            timeEntry.getOwnerId(), accrual.getAgreementId().toString(), referenceDate.plusDays(1));
+    accrualsToUpdate.add(0, accrual);
+
+    BigDecimal priorCumulativeTotal =
+        restClient.getPriorAccrual(timeEntry.getTenantId(), timeEntry.getOwnerId(),
+            accrual.getAccrualTypeId().toString(), referenceDate).getCumulativeTotal();
+
+    return updateSubsequentAccruals(accrualsToUpdate, priorCumulativeTotal);
+  }
+
+  List<Accrual> updateSubsequentAccruals(List<Accrual> accruals, BigDecimal priorCumulativeTotal) {
+
+    List<Accrual> updatedAccruals = List.copyOf(accruals);
+    //update the cumulative total for referenceDate
+    updatedAccruals.get(0).setCumulativeTotal(
+        priorCumulativeTotal.add(updatedAccruals.get(0).getContributions().getTotal()));
+
+    //cascade through until end of agreement
+    for (int i = 1; i < updatedAccruals.size(); i++) {
+      BigDecimal priorTotal =
+          updatedAccruals.get(i - 1).getCumulativeTotal();
+      Accrual currentAccrual = updatedAccruals.get(i);
+      currentAccrual.setCumulativeTotal(
+          priorTotal.add(currentAccrual.getContributions().getTotal()));
+    }
+
+    return updatedAccruals;
+  }
+
+  Accrual recalculateContributions(String timeEntryId, String tenantId,
+      String personId, AccrualType accrualType,
+      LocalDate accrualDate,
+      Range<ZonedDateTime> dateTimeRange) {
+
+    Accrual accrual =
+        restClient.getAccrualByTypeAndDate(tenantId, personId,
+            accrualType.getId().toString(), accrualDate);
+    // TODO what to do if there are no accrual records ?
```

Comment:
It's not impossible that someone records time ahead of the agreement being present. However the plan is that when an agreement is created/ changed (incl. mid year changes), time entries will be reprocessed. Therefore if we detect there is no agreement for the received time entry we can short circuit and perform no processing. This will change slightly in future when we will have to deal with time entries moving e.g. when processing a time entry change we may need to remove a previous contribution from an older agreement.

Reply:
As discussed, implemented the following logic: 
When the agreement or the accrual records can't be found stop processing, log a warning/error and return an empty list, meaning no accrual change is happening (All or Nothing)